src/managers/config.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling configuration building + writing."""
+
+import logging
+
+from ops.model import ConfigData
+
+from core.cluster import ClusterState
+from core.workload import WorkloadBase
+from literals import CONFIG_FILE
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_PROPERTIES = """
+initial-cluster-token: 'etcd-cluster'
+snapshot-count: 10000
+heartbeat-interval: 100
+election-timeout: 1000
+quota-backend-bytes: 0
+max-snapshots: 5
+max-wals: 5
+strict-reconfig-check: false
+enable-pprof: true
+proxy: 'off'
+proxy-failure-wait: 5000
+proxy-refresh-interval: 30000
+proxy-dial-timeout: 1000
+proxy-write-timeout: 5000
+proxy-read-timeout: 0
+force-new-cluster: false
+auto-compaction-mode: periodic
+auto-compaction-retention: "1"
+"""
+
+# these config properties are not used at the moment
+# they are only listed here for completeness
+TLS_PROPERTIES = """
+client-transport-security:
+  cert-file:
+  # Path to the client server TLS key file.
+  key-file:
+  client-cert-auth: false
+  trusted-ca-file:
+  auto-tls: false
+peer-transport-security:
+  cert-file:
+  key-file:
+  client-cert-auth: false
+  trusted-ca-file:
+  auto-tls: false
+  allowed-cn:
+  allowed-hostname:
+cipher-suites: [
+  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+  TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+]
+tls-min-version: 'TLS1.2'
+tls-max-version: 'TLS1.3'
+"""
+
+
+class ConfigManager:
+    """Handle the configuration of etcd."""
+
+    def __init__(
+        self,
+        state: ClusterState,
+        workload: WorkloadBase,
+        config: ConfigData,
+    ):
+        self.state = state
+        self.workload = workload
+        self.config = config
+        self.config_file = CONFIG_FILE
+
+    @property
+    def config_properties(self) -> list[str]:
+        """Assemble the config properties.
+
+        Returns:
+            List of properties to be written to the config file.
+        """
+        properties = [
+            f"name: {self.state.unit_server.member_name}",
+            f"initial-advertise-peer-urls: {self.state.unit_server.peer_url}",
+            f"initial-cluster-state: {self.state.cluster.initial_cluster_state}",
+            f"listen-peer-urls: {self.state.unit_server.peer_url}",
```

Comment:
should we not have a list of peers here?

Reply:
No not in this case. The list of peers is in `initial-cluster`. The configs `initial-advertise-peer-urls`, `listen-peer-urls`, `listen-client-urls` and `advertise-client-urls` are only for configuring the local host on which addresses to listen. Multiple urls here could be relevant for e.g. listening on traffic from localhost or listening on multiple ports. Details see [clustering guide](https://etcd.io/docs/v3.5/op-guide/clustering/#static)(scroll down to the three commands starting the three machines).